File: dotnet/GameServerSimulation/Simulation/SimulationDriver.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace GameServerSimulation.Simulation
{
    public class SimulationDriver
    {
        private readonly int _gamesCount;
        private readonly List<Reward> _boxes;
        private readonly List<AdditionalReward> _additionalRewards;

        private Dictionary<string, long> _statistics = new Dictionary<string, long>();
        public Dictionary<string, long> Statistics => _statistics;

        private Dictionary<int, long> _scoreStatistics = new Dictionary<int, long>();
        public Dictionary<int, long> ScoreStatistics => _scoreStatistics;

        public SimulationDriver(int gamesCount, List<Reward> boxes, List<AdditionalReward> additionalRewards)
        {
            _gamesCount = gamesCount;
            _boxes = boxes;
            _additionalRewards = additionalRewards;
        }

        public double Run(int interval, Action<int, double> action)
        {
            long scoresSum = 0;

            for (var i = 0; i < _gamesCount; i++)
            {
                var game = new Game(_boxes, _additionalRewards);
                var score = game.Play();
                scoresSum += score;

                var key = game.ToString();
                if (!_statistics.ContainsKey(key))
                    _statistics.Add(key, 1);
                else
                    _statistics[key] += 1;

                //

                if (!_scoreStatistics.ContainsKey(score))
                    _scoreStatistics.Add(score, 1);
                else
                    _scoreStatistics[score] += 1;

                if (i % interval == 0)
                    action(i, (double)scoresSum / i);
            }

            return (double)(scoresSum) / _gamesCount;
        }
    }
}

File: java/src/main/java/com/yggdrasilAssignment/simulation/SimulationDriver.java
package com.yggdrasilAssignment.simulation;

import com.yggdrasilAssignment.AdditionalReward;
import com.yggdrasilAssignment.Reward;

import java.util.Collection;
import java.util.Collections;
import java.util.Dictionary;
import java.util.Hashtable;

import static java.util.Collections.*;

public class SimulationDriver {
    private final int _gamesCount;
    private final Collection<Reward> _rewards;
    private final Collection<AdditionalReward> _additionalRewards;

    private final Hashtable<Integer, Long> _scoreStatistics;

    public Dictionary<Integer, Long> getScoreStatistics() { return _scoreStatistics; }

    public SimulationDriver(int gamesCount, Collection<Reward> rewards, Collection<AdditionalReward> additionalRewards) {
        _gamesCount = gamesCount;
        _rewards = unmodifiableCollection(rewards);
        _additionalRewards = unmodifiableCollection(additionalRewards);

        _scoreStatistics = new Hashtable<>();
    }

    public double Run(int interval, SimulationProgress progress) {
        long scoresSum = 0;

        for (int i = 0; i < _gamesCount; i++)
        {
            Game game = new Game(_rewards, _additionalRewards);
            int score = game.Play();
            scoresSum += score;


            if (!_scoreStatistics.containsKey(score))
                _scoreStatistics.put(score, 1L);
            else
                _scoreStatistics.put(score, _scoreStatistics.get(score) + 1);

            if (i % interval == 0)
                progress.action(i, (double)scoresSum / i);
        }

        return (double)(scoresSum) / _gamesCount;
    }
}

